test/c/test_sparse_observable.c

```diff
@@ -148,22 +179,127 @@ int test_num_qubits()
+{
+    u_int32_t num_qubits = 100;
+    SparseObservable *obs = obs_zero(num_qubits);
+
+    complex double coeff = 1;
+
+    PauliTermVec *paulis = paulis_new(); // could use with_capacity here, but we test new()
+    paulis_push(paulis, BitTerm_X, 0);
+    paulis_push(paulis, BitTerm_Y, 1);
+    paulis_push(paulis, BitTerm_Z, 2);
+
+    obs_push_copy(obs, paulis, coeff);
+    obs_push_consume(obs, paulis, coeff); // consumes the bits and indices vectors
+
+    double tol = 1e-6;
+    SparseObservable *simplified = obs_canonicalize(obs, tol);
+
+    uint64_t num_terms = obs_num_terms(obs);
+    uint64_t num_terms_simplified = obs_num_terms(simplified);
+
+    obs_deallocate(obs);
+    obs_deallocate(simplified);
+
+    if (num_terms != 2 || num_terms_simplified != 1)
+    {
+        return EqualityError;
+    }
+
+    return 0;
+}
+
+int test_term()
+{
+    SparseObservable *obs = obs_identity(100);
+
+    PauliTermVec *paulis = paulis_with_capacity(3);
+    paulis_push(paulis, BitTerm_X, 0);
+    paulis_push(paulis, BitTerm_Y, 1);
+    paulis_push(paulis, BitTerm_Z, 2);
+
+    complex double coeff = 1 + I;
+
+    obs_push_consume(obs, paulis, coeff);
+
+    // some placeholders to store the results
+    int nnis[2] = {-1, -1};
+    int bits[3] = {-1, -1, -1};
+    int indices[3] = {-1, -1, -1};
+
+    uint64_t num_terms = obs_num_terms(obs);
+    for (uint64_t i = 0; i < num_terms; i++)
+    {
+        SparseTerm *term = obs_term(obs, i);
+        uint32_t nni = obsterm_nni(term);
+        nnis[i] = nni; // store to compare later
+
+        for (uint32_t n = 0; n < nni; n++)
+        {
+            PauliTerm *pterm = obsterm_pauli(term, n);
+
+            // this loop is only called once, so we can use ``n`` to index here
+            bits[n] = pterm->bit_term;
+            indices[n] = pterm->index;
+
+            pauli_deallocate(pterm);
+        }
+
+        obsterm_deallocate(term);
+    }
+
+    obs_deallocate(obs);
+
+    int result = 0;
+    int expected_nnis[2] = {0, 3};
+    int expected_bits[3] = {BitTerm_X, BitTerm_Y, BitTerm_Z};
+    int expected_indices[3] = {0, 1, 2};
+
+    // check number of terms
+    if (num_terms != 2)
+    {
+        result = EqualityError;
+    }
+
+    // check NNIs
+    for (int i = 0; i < 2; i++)
+    {
+        if (nnis[i] != expected_nnis[i])
+        {
+            result = EqualityError;
+        }
+    }
+
+    // check bit terms and indices
+    for (int n = 0; n < 3; n++)
+    {
+        if (indices[n] != expected_indices[n] || bits[n] != expected_bits[n])
+        {
+            result = EqualityError;
+        }
+    }
+
+    return result;
+}
+
 int test_sparse_observable()
 {
     int num_failed = 0;
-
     num_failed += RUN_TEST(test_zero);
     num_failed += RUN_TEST(test_identity);
     num_failed += RUN_TEST(test_add);
-    num_failed += RUN_TEST(test_mult_real);
-    num_failed += RUN_TEST(test_mult_complex);
+    num_failed += RUN_TEST(test_mult);
+    fflush(stderr);
```

Comment:
I think you accidentally moved this line up
```suggestion
```

Reply:
yeah I was trying to get C's `printf` to print interleaved with calls to Rust's `println` but that didn't work... I'll revert 🙂